kube-runtime/src/watcher.rs

```diff
@@ -499,6 +524,110 @@ where
+                            // we have drained the last page - move on to next stage
+                            (None, State::InitPage {
+                                continue_token,
+                                objects: list.items.into_iter().collect(),
+                                last_bookmark,
+                            })
+                        }
+                        InitialListStrategy::ListWatchParallel => {
+                            // start watch
+                            match api
+                                .watch(&wc.to_watch_params(), &last_bookmark.clone().unwrap())
+                                .await
+                            {
+                                Ok(stream) => (None, State::WatchedInitPage {
+                                    continue_token,
+                                    objects: list.items.into_iter().collect(),
+                                    last_bookmark: last_bookmark,
+                                    stream,
+                                    stream_events: VecDeque::default(),
+                                }),
+                                Err(err) => {
+                                    if std::matches!(err, ClientErr::Api(ErrorResponse { code: 403, .. })) {
+                                        warn!("watch initlist error with 403: {err:?}");
+                                    } else {
+                                        debug!("watch initlist error: {err:?}");
+                                    }
+                                    (Some(Err(Error::WatchStartFailed(err))), State::Empty)
+                                }
+                            }
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                Err(err) => {
+                    if std::matches!(err, ClientErr::Api(ErrorResponse { code: 403, .. })) {
+                        warn!("watch list error with 403: {err:?}");
+                    } else {
+                        debug!("watch list error: {err:?}");
+                    }
+                    (Some(Err(Error::InitialListFailed(err))), State::Empty)
+                }
+            }
+        }
+        State::WatchedInitPage {
+            continue_token,
+            mut objects,
+            last_bookmark,
+            mut stream,
+            mut stream_events,
+        } => {
+            if let Some(next) = objects.pop_front() {
+                return (Some(Ok(Event::InitApply(next))), State::WatchedInitPage {
+                    continue_token,
+                    objects,
+                    last_bookmark,
+                    stream,
+                    stream_events,
+                });
+            }
+            // Attempt to asynchronously fetch events from the Watch Stream and cache them.
+            // If an error occurs at this stage, restart the list operation.
+            loop {
```

Comment:
This is awkward and complex; the `step` function (which were intended to do a single step), can now potentially loop for a long time.

Reply:
This step only buffers incremental events from the Watch stream into memory cache. The polling operation terminates immediately when no new events are available. Since the I/O operations are asynchronous and non-blocking, their overhead is negligible compared to the inherent latency of list operations. Therefore, this does not significantly increase the total list+watch processing time.